R/filter_raw_reads.R

```diff
@@ -22,44 +22,70 @@ filter_raw_reads = function(
-
+  cell_line_purity = sum(cell_line_filtered$n)/ sum(index_filtered$n)
+  
+  print("Generating QC table ...")
   qc_table = data.frame(cell_line_purity=cell_line_purity, index_purity = index_purity)
   
-  annotated_counts = cell_line_filtered %>%
+  # make template of expected reads
+  plate_map= sample_meta %>% dplyr::distinct(pick(c('IndexBarcode1', 'IndexBarcode2', 'pcr_plate', 'pcr_well')))
+  sample_meta$profile_id= do.call(paste,c(sample_meta[id_cols], sep=':'))
+  template= sample_meta %>% merge(cell_set_meta, by='cell_set', all.x=T) %>%
+    dplyr::mutate(members= ifelse(is.na(members), str_split(cell_set, ';'), str_split(members, ';'))) %>% 
+    unnest(cols=c(members)) %>%
+    merge(cell_line_meta, by.x= 'members', by.y= 'LUA', all.x= T)
+  
+  # check for control barcodes and add them to the template
+  if ('Y' %in% sample_meta$control_barcodes | T %in% sample_meta$control_barcodes) {
+    cb_template= sample_meta %>% dplyr::filter(control_barcodes %in% c('Y', 'T', T)) %>%
+      dplyr::mutate(joiner = 'temp') %>%
+      merge(CB_meta %>% dplyr::mutate(joiner= 'temp'), by='joiner') %>% dplyr::select(-joiner)
+    template= plyr::rbind.fill(template, cb_template)
+  }
+  
+  # annotating reads now takes much longer
+  print("Annotating reads ...")
+  annotated_counts= raw_counts %>%
+    merge(cell_line_meta, by.x="forward_read_cl_barcode", by.y="Sequence", all.x=T) %>%
     merge(CB_meta, by.x="forward_read_cl_barcode", by.y="Sequence", all.x=T) %>%
-    dplyr::select_if(function(col) sum(is.na(col)) < length(col)) %>%
-    dplyr::select(-any_of(setdiff(c("flowcell_name", "flowcell_lane", "index_1", "index_2", "members",
-                                    "lysate_well", "lysate_plate","forward_read_cl_barcode", "LUA", "pcr_well", "pcr_plate"),
-                                  id_cols))) %>%
-    dplyr::relocate(any_of(c("project_code", "CCLE_name", "DepMap_ID", "prism_cell_set", "Name", 
-                             "log2_dose", "profile_id", "trt_type", 
-                             "control_barcodes", "bio_rep", "tech_rep"))) %>%
-    dplyr::relocate(n, .after=last_col())
+    merge(plate_map, by.x= c('index_1', 'index_2'), by.y= c('IndexBarcode1', 'IndexBarcode2'), all.x=T) %>%
+    merge(template, 
+          by.x= c('index_1', 'index_2', 'forward_read_cl_barcode', intersect(colnames(template), colnames(.))), 
+          by.y= c('IndexBarcode1', 'IndexBarcode2', 'Sequence', intersect(colnames(template), colnames(.))),
+          all.x=T, all.y=T) %>% 
+    dplyr::mutate(n= replace_na(n, 0))
+  
+  # filtered counts
+  print("Filtering reads ...")
+  filt_cols= c('project_code', 'pcr_plate', 'pcr_well', 'CCLE_name', 'DepMap_ID', 'prism_cell_set',
+               'control_barcodes', 'Name', 'log2_dose','profile_id', 'trt_type')
+  filtered_counts= annotated_counts %>% dplyr::filter(!is.na(project_code)) %>%
+    dplyr::select(any_of(c(filt_cols, id_cols, 'n'))) %>%
+    dplyr::mutate(flag= ifelse(n==0, 'Missing', NA))
+  
```

Comment:
`filtered_counts` is generated from filtering annotated counts. Dependency on `project_code` column - this is expected to be filled out in every assay.